File: src/main/java/net/glasslauncher/mods/glassguis/mixin/InventoryAdditions.java
package net.glasslauncher.mods.glassguis.mixin;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.llamalad7.mixinextras.sugar.Local;
import lombok.Getter;
import lombok.Setter;
import net.fabricmc.loader.api.FabricLoader;
import net.glasslauncher.mods.gcapi3.api.CharacterUtils;
import net.glasslauncher.mods.glassguis.DrawDirection;
import net.glasslauncher.mods.glassguis.compat.AlwaysMoreItemsCompat;
import net.glasslauncher.mods.glassguis.compat.StationAPICompat;
import net.glasslauncher.mods.glassguis.screen.GlassScreen;
import net.glasslauncher.mods.glassguis.screen.widget.GlassWidget;
import net.glasslauncher.mods.glassguis.screen.widget.slot.GlassSlot;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.DrawContext;
import net.minecraft.client.gui.screen.Screen;
import net.minecraft.client.gui.screen.ingame.HandledScreen;
import net.minecraft.client.render.Tessellator;
import net.minecraft.screen.slot.Slot;
import org.lwjgl.input.Mouse;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.Shadow;
import org.spongepowered.asm.mixin.Unique;
import org.spongepowered.asm.mixin.injection.At;
import org.spongepowered.asm.mixin.injection.Inject;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfo;

import javax.imageio.ImageIO;
import java.awt.*;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import static net.glasslauncher.mods.glassguis.events.init.GlassGUIs.IMAGE_SIZE_CACHE;

@Mixin(Screen.class)
public class InventoryAdditions extends DrawContext implements GlassScreen<Screen> {
    @Unique
    public List<GlassWidget> widgets = new ArrayList<>();

    @Shadow public int width;

    @Shadow public int height;

    @Unique
    public int glassguis_guiBorder;
    @Unique
    public int glassguis_guiRoundingLight;
    @Unique
    public int glassguis_guiRoundingDark;
    @Unique
    public int glassguis_guiBackground;
    @Unique
    public int glassguis_slotRoundingLight;
    @Unique
    public int glassguis_slotRoundingDark;
    @Unique
    public int glassguis_slotBackground;
    @Unique
    public int glassguis_textColor;

    @Unique
    public String glassguis_name;

    @Inject(method = "<init>", at = @At("RETURN"))
    private void colorInit(CallbackInfo ci) {
        glassguis_guiBorder = CharacterUtils.getIntFromColour(new Color(0, 0, 0));
        glassguis_guiRoundingLight = -1;
        glassguis_guiRoundingDark = CharacterUtils.getIntFromColour(new Color(85, 85, 85));
        glassguis_guiBackground = CharacterUtils.getIntFromColour(new Color(198, 198, 198));
        glassguis_slotRoundingLight = glassguis_guiRoundingLight;
        glassguis_slotRoundingDark = CharacterUtils.getIntFromColour(new Color(55, 55, 55));
        glassguis_slotBackground = CharacterUtils.getIntFromColour(new Color(139, 139, 139));
        glassguis_textColor = CharacterUtils.getIntFromColour(new Color(63, 63, 63));
    }

    @Override
    public void glassguis_renderBackground(HandledScreen screen) {
        glassguis_renderBackground(screen.width, screen.height, screen.backgroundWidth, screen.backgroundHeight);
    }

    @Override
    public void glassguis_renderBackground(int screenWidth, int screenHeight, int backgroundWidth, int backgroundHeight) {
        int x = (screenWidth - backgroundWidth) / 2;
        int y = (screenHeight - backgroundHeight) / 2;

        glassguis_drawBackgroundBox(x, y, x + backgroundWidth, y + backgroundHeight);

        if (glassguis_name != null) {
            Minecraft.INSTANCE.textRenderer.draw(glassguis_name, x + 8, y + 6, glassguis_textColor);
        }
    }

    @Override
    public void glassguis_drawBackgroundBox(int x, int y, int x2, int y2) {
        fill(x + 2, y + 2, x2 - 2, y2 - 2, glassguis_guiBackground);

        drawHorizontalLine(x + 2, x2 - 4, y, glassguis_guiBorder); // top
        drawVerticalLine(x2 - 1, y + 2, y2 - 2, glassguis_guiBorder); // right
        drawHorizontalLine(x + 3, x2 - 3, y2 - 1, glassguis_guiBorder); // bottom
        drawVerticalLine(x, y + 1, y2 - 3, glassguis_guiBorder); // left

        glassguis_fillInclusive(x + 1, y + 1, x + 1, y + 1, glassguis_guiBorder); // tl
        glassguis_fillInclusive(x2 - 3, y + 1, x2 - 3, y + 1, glassguis_guiBorder); // tr1
        glassguis_fillInclusive(x2 - 2, y + 2, x2 - 2, y + 2, glassguis_guiBorder); // tr2
        glassguis_fillInclusive(x2 - 2, y2 - 2, x2 - 2, y2 - 2, glassguis_guiBorder); // br
        glassguis_fillInclusive(x + 2, y2 - 2, x + 2, y2 - 2, glassguis_guiBorder); // bl1
        glassguis_fillInclusive(x + 1, y2 - 3, x + 1, y2 - 3, glassguis_guiBorder); // bl2

        // light rounding
        glassguis_fillInclusive(x + 1, y + 2, x + 2, y2 - 4, glassguis_guiRoundingLight); // l
        glassguis_fillInclusive(x + 2, y + 1, x2 - 4, y + 2, glassguis_guiRoundingLight); // t
        glassguis_fillInclusive(x + 3, y + 3, x + 3, y + 3, glassguis_guiRoundingLight);

        // dark rounding
        glassguis_fillInclusive(x2 - 3, y + 3, x2 - 2, y2 - 3, glassguis_guiRoundingDark); // r
        glassguis_fillInclusive(x + 3, y2 - 3, x2 - 3, y2 - 2, glassguis_guiRoundingDark); // b
        glassguis_fillInclusive(x2 - 4, y2 - 4, x2 - 4, y2 - 4, glassguis_guiRoundingDark);
    }

    @Override
    public void glassguis_fillInclusive(int x, int y, int x2, int y2, int color) {
        fill(x, y, x2 + 1, y2 + 1, color);
    }

    @Override
    public void glassguis_drawImage(DrawContext screen, String imageString, int x, int y) {
        glassguis_drawImagePercentage(screen, imageString, x, y, 1f, DrawDirection.UP);
    }

    @Override
    public void glassguis_drawImagePercentage(DrawContext screen, String imageString, int x, int y, float percentage, DrawDirection drawDirection) {
        if (screen instanceof HandledScreen handledScreen) {
            x += ((handledScreen.width - handledScreen.backgroundWidth) / 2);
            y += ((handledScreen.height - handledScreen.backgroundHeight) / 2);
        }

        int[] size = IMAGE_SIZE_CACHE.asMap().computeIfAbsent(imageString, key -> {
            try {
                BufferedImage image = ImageIO.read(getClass().getResourceAsStream(key));
                return new int[]{image.getWidth(), image.getHeight()};
            } catch (IOException e) {
                throw new RuntimeException(e);
            }
        });

        int startX = drawDirection.left ? size[0] - (int) (size[0] * percentage) : 0;
        int startY = drawDirection.up ? size[1] - (int) (size[1] * percentage) : 0;
        x += startX;
        y += startY;

        Minecraft.INSTANCE.textureManager.bindTexture(Minecraft.INSTANCE.textureManager.getTextureId(imageString));
        glassguis_drawTexture(x, y, drawDirection.right || drawDirection.left ? (int) (size[0] * percentage) : size[0], drawDirection.down || drawDirection.up ? (int) (size[1] * percentage) : size[1], size[0], size[1], startX, startY);
    }

    @Override
    public void glassguis_drawSlots(HandledScreen screen) {
        if (screen.container == null || screen.container.slots == null) {
            return;
        }
        int x = (screen.width - screen.backgroundWidth) / 2;
        int y = (screen.height - screen.backgroundHeight) / 2;

        for (Object slotObj : screen.container.slots) {
            if (!(slotObj instanceof Slot slot)) {
                continue;
            }
            int slotX = x + slot.x;
            int slotY = y + slot.y;
            int width;
            int height;
            int backgroundWidth;
            int backgroundHeight;
            if (slot instanceof GlassSlot customSlot) {
                width = customSlot.getWidth();
                height = customSlot.getHeight();
                backgroundWidth = customSlot.getBackgroundWidth();
                backgroundHeight = customSlot.getBackgroundHeight();
            } else {
                width = 16;
                height = 16;
                backgroundWidth = 16;
                backgroundHeight = 16;
            }
            int slotBackgroundX = slotX - ((backgroundWidth - 16) / 2);
            int slotBackgroundY = slotY - ((backgroundHeight - 16) / 2);

            if (slot.getBackgroundTextureId() != -1 && ((slot instanceof GlassSlot && ((GlassSlot) slot).keepBackgroundTexture()) || !slot.hasStack())) {
                if (FabricLoader.getInstance().isModLoaded("stationapi")) {
                    StationAPICompat.drawSprite(slot, x, y, this);
                }
            }

            screen.fill(slotBackgroundX - 1, slotBackgroundY - 1, slotBackgroundX + backgroundWidth + 1, slotBackgroundY + backgroundHeight + 1, glassguis_slotBackground);
            screen.drawHorizontalLine(slotBackgroundX, slotBackgroundX + backgroundWidth - 1, slotBackgroundY - 1, glassguis_slotRoundingDark);
            screen.drawVerticalLine(slotBackgroundX - 1, slotBackgroundY - 2, slotBackgroundY + backgroundHeight, glassguis_slotRoundingDark);
            screen.drawHorizontalLine(slotBackgroundX, slotBackgroundX + backgroundWidth, slotBackgroundY + backgroundHeight, glassguis_slotRoundingLight);
            screen.drawVerticalLine(slotBackgroundX + backgroundWidth, slotBackgroundY - 1, slotBackgroundY + backgroundHeight, glassguis_slotRoundingLight);
        }
    }

    @Override
    public void glassguis_drawTexture(int x, int y, int width, int height) {
        glassguis_drawTexture(x, y, width, height, width, height, 0, 0);
    }

    @Override
    public void glassguis_drawTexture(int x, int y, int width, int height, int imgWidth, int imgHeight) {
        glassguis_drawTexture(x, y, width, height, imgWidth, imgHeight, 0, 0);
    }

    @Override
    public void glassguis_drawTexture(int x, int y, int width, int height, int imgWidth, int imgHeight, int startX, int startY) {
        double startU = (1.0 / imgWidth) * startX;
        double startV = (1.0 / imgHeight) * startY;
        double u = startU + ((1.0 / imgWidth) * width);
        double v = startV + ((1.0 / imgHeight) * height);
        Tessellator tessellator = Tessellator.INSTANCE;
        tessellator.startQuads();
        tessellator.vertex(x, y + height, 0.0, startU, v); // bl
        tessellator.vertex(x + width, y + height, 0.0, u, v); // br
        tessellator.vertex(x + width, y, 0.0, u, startV); // tr
        tessellator.vertex(x, y, 0.0, startU, startV); // tl
        tessellator.draw();
    }

    @Override
    public void glassguis_drawText(HandledScreen screen, String text, int x, int y, int color) {
        int offsetX = ((screen.width - screen.backgroundWidth) / 2) + x;
        int offsetY = ((screen.height - screen.backgroundHeight) / 2) + y;
        Minecraft.INSTANCE.textRenderer.draw(text, offsetX, offsetY, color);
    }

    @Override
    public void glassguis_tooltip(HandledScreen screen, List<String> text, Rectangle location, int mouseX, int mouseY) {
        int xOffset = (screen.width - screen.backgroundWidth) / 2;
        int yOffset = (screen.height - screen.backgroundHeight) / 2;

        if (!location.contains(mouseX - xOffset, mouseY - yOffset)) {
            return;
        }

        if (FabricLoader.getInstance().isModLoaded("alwaysmoreitems")) { // Use AMI's much better tooltip system if it's installed.
            //noinspection unchecked cry some more
            AlwaysMoreItemsCompat.setTooltip((List<Object>) (Object) text, mouseX, mouseY);
            return;
        }

        // Otherwise we're rawdogging and praying nothing else is trying to draw a tooltip.
        int startX = mouseX - xOffset + 12;
        int startY = mouseY - yOffset - 12;
        int maxLineLength = 0;
        for (String line : text) {
            if (line.length() > maxLineLength) {
                maxLineLength = line.length();
            }
        }
        if (maxLineLength == 0) {
            return;
        }
        screen.fill(
                startX - 3,
                startY - 3,
                startX + maxLineLength + 3,
                startY + 8 + 3,
                -1073741824
        );
        for (String line : text) {
            Minecraft.INSTANCE.textRenderer.drawWithShadow(line, startX, startY, -1);
            startY += 12;
        }
    }

    @Override
    public Screen glassguis_getReal() {
        return GlassScreen.super.glassguis_getReal();
    }

    @Override
    public void glassguis_mouseScrolled(int mouseX, int mouseY, int deltaWheel) {
        widgets.forEach(e -> e.onMouseScroll(mouseX, mouseY, deltaWheel));
    }

    @Override
    public List<GlassWidget> glassguis_getWidgets() {
        return widgets;
    }

    @Override
    public void glassguis_addWidget(GlassWidget widget) {
        widgets.add(widget);
    }

    @Inject(method = "onMouseEvent", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screen/Screen;mouseClicked(III)V"))
    public void onMouseDown(CallbackInfo ci, @Local(index = 1) int mouseX, @Local(index = 2) int mouseY) {
        widgets.forEach(e -> {
            if (e.getBounds().contains(mouseX, mouseY)) {
                e.onMouseDown(mouseX, mouseY, Mouse.getEventButton());
            }
        });
    }

    @Inject(method = "onMouseEvent", at = @At(value = "INVOKE", target = "Lnet/minecraft/client/gui/screen/Screen;mouseReleased(III)V"))
    public void onMouseUp(CallbackInfo ci, @Local(index = 1) int mouseX, @Local(index = 2) int mouseY) {
        widgets.forEach(e -> e.onMouseUp(mouseX, mouseY, Mouse.getEventButton()));
    }

    @Inject(method = "render", at = @At("TAIL"))
    public void renderWidgets(int mouseX, int mouseY, float delta, CallbackInfo ci) {
        widgets.forEach(e -> e.render(mouseX, mouseY, delta));

    }

    public int glassguis_getGuiBorder() {
        return glassguis_guiBorder;
    }

    public void glassguis_setGuiBorder(int glassguis_guiBorder) {
        this.glassguis_guiBorder = glassguis_guiBorder;
    }

    public int glassguis_getGuiRoundingLight() {
        return glassguis_guiRoundingLight;
    }

    public void glassguis_setGuiRoundingLight(int glassguis_guiRoundingLight) {
        this.glassguis_guiRoundingLight = glassguis_guiRoundingLight;
    }

    public int glassguis_getGuiRoundingDark() {
        return glassguis_guiRoundingDark;
    }

    public void glassguis_setGuiRoundingDark(int glassguis_guiRoundingDark) {
        this.glassguis_guiRoundingDark = glassguis_guiRoundingDark;
    }

    public int glassguis_getGuiBackground() {
        return glassguis_guiBackground;
    }

    public void glassguis_setGuiBackground(int glassguis_guiBackground) {
        this.glassguis_guiBackground = glassguis_guiBackground;
    }

    public int glassguis_getSlotRoundingLight() {
        return glassguis_slotRoundingLight;
    }

    public void glassguis_setSlotRoundingLight(int glassguis_slotRoundingLight) {
        this.glassguis_slotRoundingLight = glassguis_slotRoundingLight;
    }

    public int glassguis_getSlotRoundingDark() {
        return glassguis_slotRoundingDark;
    }

    public void glassguis_setSlotRoundingDark(int glassguis_slotRoundingDark) {
        this.glassguis_slotRoundingDark = glassguis_slotRoundingDark;
    }

    public int glassguis_getSlotBackground() {
        return glassguis_slotBackground;
    }

    public void glassguis_setSlotBackground(int glassguis_slotBackground) {
        this.glassguis_slotBackground = glassguis_slotBackground;
    }

    public int glassguis_getTextColor() {
        return glassguis_textColor;
    }

    public void glassguis_setTextColor(int glassguis_textColor) {
        this.glassguis_textColor = glassguis_textColor;
    }

    public String glassguis_getName() {
        return glassguis_name;
    }

    public void glassguis_setName(String glassguis_name) {
        this.glassguis_name = glassguis_name;
    }
}

File: src/main/kotlin/net/glasslauncher/mods/glassguis/InventoryAdditions.kt
package net.glasslauncher.mods.glassguis

import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import net.fabricmc.api.EnvType
import net.fabricmc.loader.api.FabricLoader
import net.glasslauncher.mods.alwaysmoreitems.gui.Tooltip
import net.glasslauncher.mods.gcapi3.api.CharacterUtils
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.DrawContext
import net.minecraft.client.gui.screen.Screen
import net.minecraft.client.gui.screen.ingame.HandledScreen
import net.minecraft.client.render.Tessellator
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.screen.ScreenHandler
import net.minecraft.screen.slot.Slot
import net.modificationstation.stationapi.api.client.texture.atlas.Atlases
import org.lwjgl.util.Rectangle
import java.awt.Color
import javax.imageio.ImageIO
import kotlin.math.abs

class InventoryAdditions(val width: Int = 176, val height: Int = 166) {
    companion object {
        val IMAGE_SIZE_CACHE: Cache<String, Array<Int>> = Caffeine.newBuilder().softValues().build()

        val GUI_BORDER = CharacterUtils.getIntFromColour(Color(0, 0, 0))
        val GUI_ROUNDING_LIGHT = -1
        val GUI_ROUNDING_DARK = CharacterUtils.getIntFromColour(Color(85, 85, 85))
        val GUI_BACKGROUND = CharacterUtils.getIntFromColour(Color(198, 198, 198))
        val SLOT_ROUNDING_LIGHT = GUI_ROUNDING_LIGHT
        val SLOT_ROUNDING_DARK = CharacterUtils.getIntFromColour(Color(55, 55, 55))
        val SLOT_BACKGROUND = CharacterUtils.getIntFromColour(Color(139, 139, 139))
        val TEXT_COLOR = CharacterUtils.getIntFromColour(Color(63, 63, 63))
    }

    var name: String? = null

    fun renderBackground(screen: HandledScreen) {
        val x: Int = (screen.width - screen.backgroundWidth) / 2
        val y: Int = (screen.height - screen.backgroundHeight) / 2

        drawBackgroundBox(screen, x, y, x + width, y + height)

        if (name != null) {
            Minecraft.INSTANCE.textRenderer.draw(name, x + 8, y + 6, TEXT_COLOR)
        }
    }

    fun drawBackgroundBox(screen: Screen, x: Int, y: Int, x2: Int, y2: Int) {
        screen.fill(x + 2, y + 2, x2 - 2, y2 - 2, GUI_BACKGROUND)

        screen.drawHorizontalLine(x + 2, x2 - 4, y, GUI_BORDER) // top
        screen.drawVerticalLine(x2 - 1, y + 2, y2 - 2, GUI_BORDER) // right
        screen.drawHorizontalLine(x + 3, x2 - 3, y2 - 1, GUI_BORDER) // bottom
        screen.drawVerticalLine(x, y + 1, y2 - 3, GUI_BORDER) // left

        screen.fillInclusive(x + 1, y + 1, x + 1, y + 1, GUI_BORDER) // tl
        screen.fillInclusive(x2 - 3,  y + 1, x2 - 3, y + 1, GUI_BORDER) // tr1
        screen.fillInclusive(x2 - 2,  y + 2, x2 - 2, y + 2, GUI_BORDER) // tr2
        screen.fillInclusive(x2 - 2, y2 - 2, x2 - 2, y2 - 2, GUI_BORDER) // br
        screen.fillInclusive(x + 2, y2 - 2, x + 2, y2 - 2, GUI_BORDER) // bl1
        screen.fillInclusive(x + 1, y2 - 3, x + 1, y2 - 3, GUI_BORDER) // bl2

        // light rounding
        screen.fillInclusive(x + 1, y + 2, x + 2, y2 - 4, GUI_ROUNDING_LIGHT) // l
        screen.fillInclusive(x + 2, y + 1, x2 - 4, y + 2, GUI_ROUNDING_LIGHT) // t
        screen.fillInclusive(x + 3, y + 3, x + 3, y + 3, GUI_ROUNDING_LIGHT)

        // dark rounding
        screen.fillInclusive(x2 - 3, y + 3, x2 - 2, y2 - 3, GUI_ROUNDING_DARK) // r
        screen.fillInclusive(x + 3, y2 - 3, x2 - 3, y2 - 2, GUI_ROUNDING_DARK) // b
        screen.fillInclusive(x2 - 4, y2 - 4, x2 - 4, y2 - 4, GUI_ROUNDING_DARK)
    }

    fun drawImage(screen: DrawContext, imageString: String, x: Int, y: Int) {
        drawImage(screen, imageString, x, y, 1f)
    }

    fun drawImage(screen: DrawContext, imageString: String, x: Int, y: Int, percentage: Float) {
        val baseX: Int
        val baseY: Int
        if (screen is HandledScreen) {
            baseX = x + ((screen.width - screen.backgroundWidth) / 2)
            baseY = y + ((screen.height - screen.backgroundHeight) / 2)
        }
        else {
            baseX = x
            baseY = y
        }

        val size: Array<Int> = IMAGE_SIZE_CACHE.get(imageString) {
            val image = ImageIO.read(javaClass.getResourceAsStream(imageString))
            return@get arrayOf(image.width, image.height)
        }

        Minecraft.INSTANCE.textureManager.bindTexture(Minecraft.INSTANCE.textureManager.getTextureId(imageString))
        drawTexture(baseX, baseY, abs(size[0] * percentage).toInt(), size[1], imgWidth = size[0])
    }

    fun drawSlots(screen: HandledScreen) {
        if (screen.container == null || screen.container.slots == null) {
            return
        }
        val x: Int = (screen.width - screen.backgroundWidth) / 2
        val y: Int = (screen.height - screen.backgroundHeight) / 2

        for (slot in screen.container.slots) {
            if (slot !is Slot) {
                continue
            }
            val slotX = x + slot.x
            val slotY = y + slot.y
            val width: Int
            val height: Int
            val backgroundWidth: Int
            val backgroundHeight: Int
            if (slot is CustomSizeSlot) {
                width = slot.getWidth()
                height = slot.getHeight()
                backgroundWidth = slot.getBackgroundWidth()
                backgroundHeight = slot.getBackgroundHeight()
            } else {
                width = 16
                height = 16
                backgroundWidth = 16
                backgroundHeight = 16
            }
            val slotBackgroundX = slotX - ((backgroundWidth - 16) / 2)
            val slotBackgroundY = slotY - ((backgroundHeight - 16) / 2)

            if (slot.backgroundTextureId != -1 && ((slot is CustomSizeSlot && slot.keepBackgroundTexture()) || !slot.hasStack())) {
                val sprite = Atlases.getGuiItems().getTexture(slot.backgroundTextureId)
                Minecraft.INSTANCE.textureManager.bindTexture(Minecraft.INSTANCE.textureManager.getTextureId("terrain.png"))
                drawTexture(x + slot.x, y + slot.y, sprite.width, sprite.height)
            }

            screen.fill(slotBackgroundX - 1, slotBackgroundY - 1, slotBackgroundX + backgroundWidth + 1, slotBackgroundY + backgroundHeight + 1, SLOT_BACKGROUND)
            screen.drawHorizontalLine(slotBackgroundX, slotBackgroundX + backgroundWidth - 1, slotBackgroundY - 1, SLOT_ROUNDING_DARK)
            screen.drawVerticalLine(slotBackgroundX - 1, slotBackgroundY - 2, slotBackgroundY + backgroundHeight, SLOT_ROUNDING_DARK)
            screen.drawHorizontalLine(slotBackgroundX, slotBackgroundX + backgroundWidth, slotBackgroundY + backgroundHeight, SLOT_ROUNDING_LIGHT)
            screen.drawVerticalLine(slotBackgroundX + backgroundWidth, slotBackgroundY - 1, slotBackgroundY + backgroundHeight, SLOT_ROUNDING_LIGHT)


        }
    }

    fun setupPlayerInventory(handledScreen: HandledScreen, playerInventory: PlayerInventory) {
        // 162 81
        val x = ((handledScreen.width - handledScreen.backgroundWidth) / 2) + (handledScreen.backgroundWidth / 2) + 8
        val y = (handledScreen.height / 2) + handledScreen.backgroundHeight + 1

        setupPlayerInventory(x, y, playerInventory, handledScreen.container)
    }

    fun setupPlayerInventory(x: Int, y: Int, playerInventory: PlayerInventory, screenHandler: ScreenHandler) {
        for (row in 0 until 3) {
            for (col in 0 until 9) {
                screenHandler.addSlot(
                    Slot(playerInventory, col + row * 9 + 9, x + col * 18, y - (4 - row) * 18 - 11)
                )
            }
        }

        for (hotbarSlot in 0 until 9) {
            screenHandler.addSlot(Slot(playerInventory, hotbarSlot, x + hotbarSlot * 18, y - 25))
        }
    }

    fun setupPlayerInventory(screenHandler: ScreenHandler, playerInventory: PlayerInventory) {
        if (FabricLoader.getInstance().environmentType == EnvType.SERVER) {
            setupPlayerInventory(0, 0, playerInventory, screenHandler)
        }
    }

    fun drawTexture(x: Int, y: Int, width: Int, height: Int, imgWidth: Int = width, imgHeight: Int = height, startX: Int = 0, startY: Int = 0) {
        val startU = (1.0 / imgWidth) * startX
        val startV = (1.0 / imgHeight) * startY
        val u = (1.0 / imgWidth) * width
        val v = (1.0 / imgHeight) * height
        val tessellator = Tessellator.INSTANCE
        tessellator.startQuads()
        tessellator.vertex((x + 0).toDouble(), (y + height).toDouble(), 0.0, startU, v) // bl
        tessellator.vertex((x + width).toDouble(), (y + height).toDouble(), 0.0, u, v) // br
        tessellator.vertex((x + width).toDouble(), (y + 0).toDouble(), 0.0, u, startV) // tr
        tessellator.vertex((x + 0).toDouble(), (y + 0).toDouble(), 0.0, startU, startV) // tl
        tessellator.draw()
    }

    fun drawText(screen: HandledScreen, text: String, x: Int, y: Int, color: Int) {
        val offsetX = ((screen.width - screen.backgroundWidth) / 2) + x
        val offsetY = ((screen.height - screen.backgroundHeight) / 2) + y
        Minecraft.INSTANCE.textRenderer.draw(text, offsetX, offsetY, color)
    }

    fun tooltip(screen: HandledScreen, text: List<String>, location: Rectangle, mouseX: Int, mouseY: Int) {
        val xOffset = (screen.width - screen.backgroundWidth) / 2
        val yOffset = (screen.height - screen.backgroundHeight) / 2

        if (!location.contains(mouseX - xOffset, mouseY - yOffset)) {
            return
        }

        if (FabricLoader.getInstance().isModLoaded("alwaysmoreitems")) { // Use AMI's much better tooltip system if it's installed.
            Tooltip.INSTANCE.setTooltip(text, mouseX, mouseY)
            return
        }

        // Otherwise we're rawdogging and praying nothing else is trying to draw a tooltip.
        val startX: Int = mouseX - xOffset + 12
        var startY: Int = mouseY - yOffset - 12
        var maxLineLength = 0
        for (line in text) {
            if (line.length > maxLineLength) {
                maxLineLength = line.length
            }
        }
        if (maxLineLength == 0) {
            return
        }
        screen.fill(
            startX - 3,
            startY - 3,
            startX + maxLineLength + 3,
            startY + 8 + 3,
            -1073741824
        )
        for (line in text) {
            Minecraft.INSTANCE.textRenderer.drawWithShadow(line, startX, startY, -1)
            startY += 12
        }
    }
}